### ISSUE TITLE
Codebook Update with Cosine Similarity

### ISSUE THREAD

**Issue by goldbird5** [lucidrains/vector-quantize-pytorch#48]
In CosineSimCodebook forward function [line#449](https://github.com/lucidrains/vector-quantize-pytorch/blob/500db140ceb2939a957e66dcf22849e40ff0daa3/vector_quantize_pytorch/vector_quantize_pytorch.py#LL449C17-L449C17), is there any reason why we use 'bins' directly instead of 'self.cluster_size'?
It seems EuclideanCodebook uses laplace smoothed 'self.cluster_size' as [line#308](https://github.com/lucidrains/vector-quantize-pytorch/blob/500db140ceb2939a957e66dcf22849e40ff0daa3/vector_quantize_pytorch/vector_quantize_pytorch.py#LL308C1-L308C1), but CosineSimCodebook doesn't.
There exists process to update 'self.cluster_size' in [line#441](https://github.com/lucidrains/vector-quantize-pytorch/blob/500db140ceb2939a957e66dcf22849e40ff0daa3/vector_quantize_pytorch/vector_quantize_pytorch.py#L441), but it seems we never use it in CosineSimCodebook class(except expiring).

**Comment by lucidrains** [comment]
@goldbird5 i don't remember why! maybe i read something from the paper that introduced it (vit vqgan) and they did not have this?

**Comment by lucidrains** [comment]
it seems like i should offer it, even if [some papers](https://openreview.net/forum?id=GxjCYmQAody) have shown that a simplified EMA (with no commitment loss nor tracking of cluster sizes) is sufficient

**Comment by lucidrains** [comment]
ohh i think i remember now, looking at the code

so when i was building the cosine sim variant, i was concurrently building n-grammer, and in their paper they said a lot of the stuff in the original quantization logic was not necessary. so therefore i did not keep track of either the embedding sum nor the cluster sizes.

however, later on, when building residual VQ, the feature for expiring codes came up, and i realized that the cluster size had to be kept track of for both libraries, and added it

**Comment by lucidrains** [comment]
![Screenshot from 2023-05-19 14-12-16](https://github.com/lucidrains/vector-quantize-pytorch/assets/108653/ea885e46-1dea-429e-8faa-8492bfdc3441)

**Comment by lucidrains** [comment]
@goldbird5 let me know if [this](https://github.com/lucidrains/vector-quantize-pytorch/commit/03da0092d35f2205612a37ca6dddf2900f6c1ba0) looks better

if somehow it was better the way before, we can roll back

**Comment by lucidrains** [comment]
@goldbird5 while looking at the code, think i also uncovered a bug with multi-headed codebooks and the cluster size smoothing 🤦‍♂️

**Comment by goldbird5** [comment]
@lucidrains Thank you for checking! Good to hear that you uncovered a new bug
I still wonder why there are slight differences in updating between Euclidean and CosineSim.
In Euclidean, it shows to directly copy 'embed_normalized' to 'self.embed',
but in CosineSim, there is one more step which is applying EMA on 'self.embed' using l2-normalized 'embed_normalized', and copy itself with l2-normalizing([here](https://github.com/lucidrains/vector-quantize-pytorch/blob/afa227589b2aca82db4c4773513e36176428b4b2/vector_quantize_pytorch/vector_quantize_pytorch.py#LL456C40-L456C40).

Is there any reason? or is it rather to be modified to be consistent in both classes?
I don't know which way is better though.

**Comment by lucidrains** [comment]
@goldbird5 oh no you are right, made the fix https://github.com/lucidrains/vector-quantize-pytorch/commit/a7bc53b1f0297e55eca0c1c44dce126ccc3feab3

**Comment by goldbird5** [comment]
Thanks again!

**Comment by lucidrains** [comment]
@goldbird5 no thank you! def let me know how this fares during training
